presto-main/src/main/java/com/facebook/presto/server/DynamicFilterService.java

```diff
@@ -0,0 +1,207 @@
+    }
+
+    public synchronized void removeQuery(String queryId)
+    {
+        dynamicFilterSummaries.entrySet().removeIf(entry -> entry.getKey().getQueryId().equals(queryId));
+    }
+
+    @Immutable
+    private static class SourceDescriptor
+    {
+        private final String queryId;
+        private final String source;
+
+        public static SourceDescriptor of(String queryId, String source)
+        {
+            return new SourceDescriptor(queryId, source);
+        }
+
+        private SourceDescriptor(String queryId, String source)
+        {
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.source = requireNonNull(source, "source is null");
+        }
+
+        public String getQueryId()
+        {
+            return queryId;
+        }
+
+        public String getSource()
+        {
+            return source;
+        }
+
+        @Override
+        public boolean equals(Object other)
+        {
+            if (other == this) {
+                return true;
+            }
+            if (other == null || !(other instanceof SourceDescriptor)) {
+                return false;
+            }
+
+            SourceDescriptor sourceDescriptor = (SourceDescriptor) other;
+
+            return Objects.equals(queryId, sourceDescriptor.queryId) &&
+                    Objects.equals(source, sourceDescriptor.source);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(queryId, source);
+        }
+    }
+
+    @NotThreadSafe
+    public static class DynamicFilterSummaryWithSenders
+    {
+        // Key in form of "stageId.taskId"
+        private final Map<String, SenderStats> senderStats = new HashMap<>();
+        private DynamicFilterSummary dynamicFilterSummary;
+
+        public Optional<DynamicFilterSummary> getSummaryIfReady()
+        {
+            for (Map.Entry<String, SenderStats> entry : senderStats.entrySet()) {
+                if (!entry.getValue().isCompleted()) {
+                    return Optional.empty();
+                }
+            }
+
+            return Optional.of(dynamicFilterSummary);
+        }
+
+        public void registerTask(int stageId, int taskId)
+        {
+            String stageTaskId = stageId + "." + taskId;
+            senderStats.put(stageTaskId, new SenderStats());
+        }
+
+        public void addSummary(int stageId, int taskId, int driverId, DynamicFilterSummary summary, int expectedSummariesCount)
+        {
+            String stageTaskId = stageId + "." + taskId;
+
+            SenderStats stats = senderStats.get(stageTaskId);
+            if (stats == null || !stats.isValid()) {
+                stats = new SenderStats(expectedSummariesCount);
+            }
+            stats.addSummary(driverId);
+            // override the stats, in case it's first summary delivered (task registered but first summary not yet delivered)
+            senderStats.put(stageTaskId, stats);
```

Comment:
Put overrides the object if already exists. This should not be a problem here. But I'll change it.